website/www/site/content/en/documentation/ml/anomaly-detection.md

```diff
@@ -0,0 +1,238 @@
+
+The file structure for anomaly_detection pipeline is:
+
+    anomaly_detection_pipeline/
+    ├── pipeline/
+    │   ├── __init__.py
+    │   ├── options.py
+    │   └── transformations.py
+    ├── __init__.py
+    ├── config.py
+    ├── main.py
+    └── setup.py
+
+`pipeline/transformations.py` contains the code for different `beam.DoFn` and additional functions that are used in pipeline
+
+`pipeline/options.py` contains the pipeline options to configure the Dataflow pipeline
+
+`config.py` defines some variables like GCP PROJECT_ID, NUM_WORKERS that are used multiple times
+
+`setup.py` defines the packages/requirements for the pipeline to run
+
+`main.py` contains the pipeline code and some additional functions used for running the pipeline
+
+### How to Run the Pipeline ?
+First, make sure you have installed the required packages and you have pushed data to PubSub.
+
+1. Locally on your machine: `python main.py`
+2. On GCP for Dataflow: `python main.py --mode cloud`
+
+The pipeline can be broken down into few simple steps:
+
+1. Reading the message from PubSub
+2. Converting the PubSub message into a PCollection of dictionaries where the key is the UID and the value is the twitter text
+3. Encoding the text into transformer-readable token ID integers using a tokenizer
+4. Using RunInference to get the vector embedding from a Transformer based Language Model
+5. Normalizing the embedding
+6. Using RunInference to get anomaly prediction from a trained HDBSCAN clustering model
+7. Writing the prediction to BQ, so that clustering model can be retrained when needed
+8. Sending an email alert if anomaly is detected
+
+
+The code snippet for the first two steps of the pipeline:
+
+{{< highlight >}}
+    docs = (
+        pipeline
+        | "Read from PubSub"
+        >> ReadFromPubSub(subscription=cfg.SUBSCRIPTION_ID, with_attributes=True)
+        | "Decode PubSubMessage" >> beam.ParDo(Decode())
+    )
+{{< /highlight >}}
+
+We will now focus on important steps of pipeline: tokenizing the text, getting embedding using RunInference and finally getting prediction from HDBSCAN model.
+
+### Getting Embedding from a Language Model
+
+In order to do clustering with text data, we first need to map the text into vectors of numerical values suitable for statistical analysis. We use a transformer based language model called [sentence-transformers/stsb-distilbert-base/stsb-distilbert-base](https://huggingface.co/sentence-transformers/stsb-distilbert-base). It maps sentences & paragraphs to a 768 dimensional dense vector space and can be used for tasks like clustering or semantic search. But, we first need to tokenize the text as the language model is expecting a tokenized input instead of raw text.
+
+Tokenization is a preprocessing task that transforms text in a way that it can be fed into the model for getting predictions.
```

Comment:
```suggestion
Tokenization is a preprocessing task that transforms text so that it can be fed into the model for getting predictions.
```

Sorry, one more wording nit I missed last time